UI - Add hop creation and peers mgmt

tanganellilore opened the pull request:
##### SUMMARY
As discussed with @TheRealHaoLiu and @fosterseth this PR is to add new features on UI for
- Modify add new instace UI witch checkbox for new param `peers_from_control_nodes` and multi popup choice for peers
- usage of `/api/v2/instances` patch for manage peers
- add associate/disassociate buttons on instance peers tab (visible only for hop and execution node)
- add edit page for instance with possibility of change description, port, enablement and peer to control node

Below some screenshots:

## New Add Instance UI
![image](https://user-images.githubusercontent.com/35271287/234941782-95d14d72-ba70-43f3-90f6-bc74d4b76cc9.png)

#### Choice peers
![image](https://user-images.githubusercontent.com/35271287/234942404-dd5f776d-0c0a-4ab0-994c-ea701875d325.png)

#### UI with peer selected
![image](https://user-images.githubusercontent.com/35271287/234942483-4424d291-58f6-4116-8cc7-5f9178c2a418.png)

## New peers add/deassociate button
![image](https://user-images.githubusercontent.com/35271287/234942677-8c1c497b-e200-4835-99cf-08d107ab4da0.png)

#### when you click on `Add ` (similar to add instance). If you click save peers will be added
![image](https://user-images.githubusercontent.com/35271287/234942853-85609381-375f-421f-a513-a8faab3577f1.png)

#### If you want to remove, you can simply select and click on `Disassociate`. Popup warning will be showed
![image](https://user-images.githubusercontent.com/35271287/234943033-e466a7e2-5b57-4337-9b98-1b839546c632.png)


## Edit Page

#### Details page have a `Edit` button
![image](https://user-images.githubusercontent.com/35271287/234943191-6b40ce57-1918-4319-b362-80aefa9fc995.png)

#### if you click on `Edit`, a similar view of adding page will be prompted. Served on `/instances/:id/edit`
![image](https://user-images.githubusercontent.com/35271287/234943322-06cac1c8-3921-412d-99be-507ee3a6c217.png)

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
 - UI

##### AWX VERSION

```
N/A
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
add UI for mgmt peers, instance edit and add

Commit:
fix UI lint and test screen

fosterseth commented:
![image](https://user-images.githubusercontent.com/8745776/234974382-86969a7b-8ab1-4e4e-a4da-29f72540f22f.png)
would be nice to show (read only) the "connect to control nodes" here. I do see it on the edit page

Commit:
add peer info on detail and bug fix on detail

tanganellilore commented:
added on last commit.

## Enabled:
![image](https://user-images.githubusercontent.com/35271287/234985783-a01afb5b-a655-486e-92e2-abb5e4d6fc67.png)

## Disabled:
![image](https://user-images.githubusercontent.com/35271287/234985978-2983c819-6dd9-4a50-99ba-d65cc0e738ad.png)

I also fixed one bug on UI for checkbox

Commit:
fix lint

kialam commented:
This is great work - thank you for submitting this for review! Some initial notes:
1. Boolean fields are represented in the UI as label and label detail components. See example:
![Screenshot 2023-05-01 at 10 17 05 AM](https://user-images.githubusercontent.com/2293210/235495695-ddb40ded-29eb-4cca-8be2-2d11aafb8528.png)
2. Fields that are not editable in a form should be greyed out - e.g. the node type on the Instance edit form should not be a selectable choice.
![Screenshot 2023-05-01 at 10 24 55 AM](https://user-images.githubusercontent.com/2293210/235496740-871c9b0c-b363-4a78-aa26-b5d92b639d95.png)
3. The `Add` button for peers list should be called `Associate` to keep in line with the `Disassociate` button. We have the same model for Instance Groups and Instances.
![Screenshot 2023-05-01 at 10 25 11 AM](https://user-images.githubusercontent.com/2293210/235496766-f9694e0f-4354-47e0-b76a-083f38015abf.png)
5. The Edit button should always be anchored on the left and the primary button.
![Screenshot 2023-05-01 at 10 25 27 AM](https://user-images.githubusercontent.com/2293210/235496783-22b4136d-8fd1-455e-a3b5-d72d22445fd0.png)
6. Option fields should be grouped together like so:
![Screenshot 2023-05-01 at 10 28 33 AM](https://user-images.githubusercontent.com/2293210/235497186-0e5de840-711c-4808-9745-d2113bf027b8.png)

tanganellilore commented:
@kialam Thanks for reviews. I will answer per points:
1. Ok i will try to modify it (i used a chip in the detail view to be more "modern" but if this is not allowed il will remove it)
2. As discusse with @TheRealHaoLiu should be possibile change the type of node after creation, this is why was enable the modify. Let me know which modification it's required.
3. Agree and also I suppose to change add with associate.
4. I will do it
5. I will do the grouping

Tomorrow i will add one commit for 1,3,4,5.
For point 2 i will wait your feedback

kialam commented:
@tanganellilore thank you. I spoke with @fosterseth and @TheRealHaoLiu and we want to disable changing node type in the edit form for now.

tanganellilore commented:
Ok, no problem.
Tomorrow i will send new commit on this branch

kialam commented:
Looking at the failed unit test: 
`InstancesLookup › should render successfully with instance_details for edit instance` - I don't see an Instances look up the edit form: 
```
{!isEdit && (
        <InstancesLookup
          helperTextInvalid={peersMeta.error}
          isValid={!peersMeta.touched || !peersMeta.error}
          ...
```

tanganellilore commented:
Form is one for add and edit.
In edit part I removed the peers because are managed on peers tab.

Anyway, I will add it with using peers api of instance.

Commit:
fix UI and ui-test

tanganellilore commented:
@kialam new commit submitted with all fixing discussed. Please re-run test to check if something is missing

kialam commented on awx/ui/src/components/Lookup/InstancesLookup.js:
```diff
@@ -0,0 +1,212 @@
+  order_by: 'hostname',
+});
+
+function InstancesLookup({
```
I'm debating if this is better named PeersLookup. The merged query params and filtering make this a very specific subset of Instances being queried and is not merely a generic Instances Lookup.

kialam commented on awx/ui/src/components/StatusIcon/icons.js:
```diff
@@ -41,6 +41,7 @@ const icons = {
   skipped: MinusCircleIcon,
   canceled: ExclamationTriangleIcon,
   changed: ExclamationTriangleIcon,
+  enabled: CheckCircleIcon,
```
We can remove this

kialam commented on awx/ui/src/components/StatusLabel/StatusLabel.js:
```diff
@@ -24,6 +24,7 @@ const colors = {
   disabled: 'grey',
   canceled: 'orange',
   changed: 'orange',
+  enabled: 'green',
```
Same here

kialam replied:
```suggestion
```

kialam commented on awx/ui/src/screens/Instances/InstanceDetail/InstanceDetail.js:
```diff
@@ -205,13 +206,41 @@ function InstanceDetail({ setBreadcrumb, isK8s }) {
+            <Detail
+              label={t`Connection to control nodes`}
+              value={
+                instance.peers_from_control_nodes ? t`Enabled` : t`Disabled`
```
```suggestion
                instance.peers_from_control_nodes ? t`On` : t`Off`
```

kialam commented on awx/ui/src/components/StatusLabel/StatusLabel.js:
```diff
@@ -53,6 +54,7 @@ export default function StatusLabel({ status, tooltipContent = '', children }) {
     disabled: t`Disabled`,
     canceled: t`Canceled`,
     changed: t`Changed`,
+    enabled: t`Enabled`,
```
```suggestion
```

kialam commented on awx/ui/src/components/StatusLabel/StatusLabel.js:
```diff
@@ -104,6 +106,7 @@ StatusLabel.propTypes = {
     'disabled',
     'canceled',
     'changed',
+    'enabled',
```
```suggestion
```

Commit:
remove unused chip and change peer label

tanganellilore commented:
@kialam removed unused lines and changed the label for peers.

Only open point for "instancelookup" but comment is above.

kialam commented on awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js:
```diff
@@ -0,0 +1,143 @@
+  let wrapper;
+  let history;
+
+  beforeAll(async () => {
```
```suggestion
  beforeEach(async () => {
```
^ fix the failing unit tests

kialam commented:
The last unit test appears to have caught a bug:
```
<ErrorDetail__CardBody>
                                            <PFCardBody className="ErrorDetail__CardBody-sc-1grrpky-1 eUlxDj">
                                              <div className="pf-c-card__body ErrorDetail__CardBody-sc-1grrpky-1 eUlxDj">
                                                <strong>
                                                  TypeError
                                                  : 
                                                  Cannot read properties of undefined (reading &#39;data&#39;)
                                                </strong>
                                              </div>
                                            </PFCardBody>
                                          </ErrorDetail__CardBody>
                                          <ErrorDetail___StyledCardBody2>
                                            <PFCardBody className="ErrorDetail__CardBody-sc-1grrpky-1 ErrorDetail___StyledCardBody2-sc-1grrpky-2 eUlxDj cNBfAP">
                                              <div className="pf-c-card__body ErrorDetail__CardBody-sc-1grrpky-1 ErrorDetail___StyledCardBody2-sc-1grrpky-2 eUlxDj cNBfAP">
                                                TypeError: Cannot read properties of undefined (reading &#39;data&#39;)
                                                    at /Users/kilam/Sites/awx/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.js:40:49
                                                    at processTicksAndRejections (node:internal/process/task_queues:96:5)
                                                    at /Users/kilam/Sites/awx/awx/ui/src/hooks/useRequest.js:32:28
                                              </div>
                                            </PFCardBody>
                                          </ErrorDetail___StyledCardBody2>
```

kialam commented:
@tanganellilore trying to get your PR to the finish line. We have some unit test failures that need to be fixed, otherwise it's looking really good. The last bit is to just change `InstancesLookup` to `PeersLookup` and we should be ready to merge.

kialam commented:
This should get the unit tests passing:
```diff
diff --git a/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js b/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js
index 64e2967f77..1447f35d61 100644
--- a/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js
+++ b/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js
@@ -74,10 +74,11 @@ describe('<InstanceEdit/>', () => {
   let wrapper;
   let history;
 
-  beforeAll(async () => {
+  beforeEach(async () => {
     useDebounce.mockImplementation((fn) => fn);
     history = createMemoryHistory();
     InstancesAPI.readDetail.mockResolvedValue({ data: instanceData });
+    InstancesAPI.readPeers.mockResolvedValue({ data: [] });
 
     await act(async () => {
       wrapper = mountWithContexts(
```

kialam reviewed (approved):
Excellent work. Once tests pass this is good to merge.

fosterseth commented:
api-test, api-swagger, awx-collection, and api-scheme are expected to fail for now, as the feature branch will also fail those. I can address those later.

However the ui-lint and ui-test-screens tests should pass I think

Commit:
fix tests

Commit:
rename lookup, put Instance type disable on edit

tanganellilore commented:
Test should be fixed (excluded @fosterseth tests).
Renamed file and disabled Instance type on edit page (instead disappear it)
